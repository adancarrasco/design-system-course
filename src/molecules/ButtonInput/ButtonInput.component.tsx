import * as React from 'react'
import styled from '@emotion/styled'
import { CloseOutlined } from '@material-ui/icons'

import Button from '../../atoms/Button'
import Input from '../../atoms/Input'
import { Typography } from '../../atoms'

const cancelId = 'cancel'
const saveId = 'save'

export interface IButtonInputProps {
  children?: string
  isEditMode?: boolean
  isEditing?: boolean
  placeholder?: string
  onClick?: (e: React.MouseEvent) => void
  saveText?: string
  onSave: (inputValue: string) => void
  onBlur?: () => void
  onChange?: () => void
  onCancelClick?: (e: React.MouseEvent) => void
}

const ButtonInput: React.FC<IButtonInputProps> = ({
  children = '',
  isEditMode = false,
  placeholder,
  onClick,
  saveText,
  onSave,
  onBlur,
  onChange,
  onCancelClick,
  ...rest
}) => {
  const [isEditing, setIsEditing] = React.useState<boolean>(isEditMode)
  const [inputValue, setInputValue] = React.useState<string>(children)
  const [prevValue, setPrevValue] = React.useState<string>('')

  const save = () => {
    onSave(inputValue)
    setIsEditing(false)
  }

  const onClickHandler = (e: React.MouseEvent) => {
    if (!isEditing) {
      setPrevValue(inputValue)
      setIsEditing(true)
      onClick && onClick(e)
    }
  }

  const onActionsClick = (e: React.MouseEvent<HTMLElement>) => {
    const element = e.target as HTMLElement
    if (
      isEditing &&
      (element.tagName.toLowerCase() === 'span' ||
        element.tagName.toLowerCase() === 'div' ||
        element.getAttribute('data-testid') === saveId)
    ) {
      save()
    } else {
      setInputValue(prevValue)
      setIsEditing(false)
    }
  }

  const onBlurHandler = () => {
    if (!saveText && isEditing) {
      onBlur && onBlur()
      save()
    }
  }

  const onCancelClickHanlder = (e) => {
    onCancelClick && onCancelClick(e)
  }

  return (
    <ButtonInputStyled
      onClick={(e) => {
        onClickHandler(e)
      }}
    >
      {isEditing ? (
        <EditInputStyled {...rest}>
          <Input
            value={inputValue}
            onChange={(e: React.ChangeEvent<HTMLTextAreaElement>) => {
              setInputValue(e.target.value)
              onChange && onChange()
            }}
            onBlur={onBlurHandler}
            placeholder={placeholder}
          />
          {saveText && (
            <ActionsContainerStyled onClick={onActionsClick}>
              <Button
                data-testid={saveId}
                backgroundColor="#3ea93e"
                backgroundColorHover="#56bd56"
                colorHover="#fff"
                onClick={save}
              >
                {saveText}
              </Button>
              <Button
                data-testid={cancelId}
                backgroundColor="transparent"
                backgroundColorHover="transparent"
                color="#ababab"
                StartIcon={CloseOutlined}
                onClick={onCancelClickHanlder}
              />
            </ActionsContainerStyled>
          )}
        </EditInputStyled>
      ) : (
        <PlaceholderStyled {...rest}>
          <Typography>{inputValue.length > 0 ? inputValue : placeholder}</Typography>
        </PlaceholderStyled>
      )}
    </ButtonInputStyled>
  )
}
const ButtonInputStyled = styled.div`
  box-sizing: border-box;
  width: 100%;
`

interface IPlaceholderStyledProps {
  backgroundColor?: string
  backgroundColorHover?: string
  isEditing?: boolean
}

const PlaceholderStyled = styled.div<IPlaceholderStyledProps>`
  cursor: pointer;
  background-color: ${({ backgroundColor, isEditing }) =>
    backgroundColor && isEditing ? backgroundColor : '#f1f1f1'};
  padding: 0.5rem 0.6rem;
  min-height: 40px;
  :hover {
    background-color: ${(props) => props.backgroundColorHover ?? '#e4e4e4'};
  }
`

const EditInputStyled = styled.div``

const ActionsContainerStyled = styled.div``

export default ButtonInput
