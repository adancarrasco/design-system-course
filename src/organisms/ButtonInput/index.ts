export { default } from './ButtonInput.component'
export type { IButtonInputProps } from './ButtonInput.component'
