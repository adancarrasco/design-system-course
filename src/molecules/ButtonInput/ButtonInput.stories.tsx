import React from 'react'
import { storiesOf } from '@storybook/react'

import ButtonInput from '.'

storiesOf('Molecules|ButtonInput', module)
  .add('Default', () => (
    <div style={{ width: '400px' }}>
      <ButtonInput
        placeholder="Add a more detailed description..."
        onClick={() => console.log('Edit mode')}
        saveText="Save"
        onSave={() => console.log('Saving...')}
        onChange={() => console.log('Changing...')}
        onCancelClick={() => console.log('Cancel click')}
      ></ButtonInput>
    </div>
  ))
  .add('No buttons', () => (
    <div style={{ width: '400px' }}>
      <ButtonInput
        placeholder="Add a more detailed description..."
        onClick={() => console.log('Edit mode')}
        onSave={() => console.log('Saving...')}
        onBlur={() => console.log('Blurring')}
        onChange={() => console.log('Changing...')}
      ></ButtonInput>
    </div>
  ))
