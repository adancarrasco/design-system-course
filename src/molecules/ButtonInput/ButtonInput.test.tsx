import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { matchers } from 'jest-emotion'
import '@testing-library/jest-dom/extend-expect'

import ButtonInput, { IButtonInputProps } from '.'

expect.extend(matchers)

describe('ButtonInput Component', () => {
  const renderComponent = (props: IButtonInputProps, dataTestId: string) =>
    render(<ButtonInput data-testid={dataTestId} {...props} />)

  let props: IButtonInputProps
  beforeEach(() => {
    props = {
      placeholder: 'Add more detailed description...',
      onClick: jest.fn(),
      saveText: 'Add another card',
      onChange: jest.fn(),
      onSave: jest.fn(),
    }
  })
  it('matches snapshot and default render', () => {
    const dataTestId = 'buttonInput'
    const { getByTestId } = renderComponent(props, dataTestId)
    const container = getByTestId('buttonInput')
    expect(container).toMatchSnapshot()
    expect(container).toBeInTheDocument()
  })

  it('should enable editing', () => {
    const dataTestId = 'buttonInput'
    const { getByTestId, getByPlaceholderText, getByText } = renderComponent(props, dataTestId)
    const container = getByTestId('buttonInput')
    expect(container.childNodes.length).toBe(1)
    fireEvent.click(container)
    const textArea = getByPlaceholderText('Add more detailed description...')
    expect(textArea).toBeInTheDocument()
    const saveButton = getByText('Add another card')
    fireEvent.click(saveButton)
  })

  it('should save text in state', async () => {
    const dataTestId = 'buttonInput'
    const { getByTestId, getByPlaceholderText, getByText } = renderComponent(props, dataTestId)
    const container = getByTestId('buttonInput')
    expect(container.childNodes.length).toBe(1)
    fireEvent.click(container)
    const textArea = getByPlaceholderText('Add more detailed description...')
    expect(textArea).toBeInTheDocument()
    await userEvent.type(textArea, 'Add some text here')
    const saveButton = getByText('Add another card')
    fireEvent.click(saveButton)
    const updatedText = getByText('Add some text here')
    expect(updatedText).toBeInTheDocument()
  })

  it('should click cancel and keep previous state', async () => {
    const dataTestId = 'buttonInput'
    const { getByTestId, getByPlaceholderText, getByText } = renderComponent(props, dataTestId)
    const container = getByTestId('buttonInput')
    expect(container.childNodes.length).toBe(1)
    fireEvent.click(container)
    const textArea = getByPlaceholderText('Add more detailed description...')
    expect(textArea).toBeInTheDocument()
    await userEvent.type(textArea, 'Add some text here')
    const cancelButton = getByTestId('cancel')
    fireEvent.click(cancelButton)
    const keptText = getByText('Add more detailed description...')
    expect(keptText).toBeInTheDocument()
  })
})
