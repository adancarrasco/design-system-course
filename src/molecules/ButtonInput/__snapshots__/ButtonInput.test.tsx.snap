// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonInput Component matches snapshot and default render 1`] = `
.emotion-1 {
  cursor: pointer;
  background-color: #f1f1f1;
  padding: 0.5rem 0.6rem;
  min-height: 40px;
}

.emotion-1:hover {
  background-color: #e4e4e4;
}

.emotion-0 {
  color: #3d3d3d;
}

<div
  class="emotion-1"
  data-testid="buttonInput"
>
  <span
    class="emotion-0"
    color="#3d3d3d"
  >
    Add more detailed description...
  </span>
</div>
`;
